eta/core/datasets.py

```diff
@@ -1370,19 +1373,294 @@ def transform(self, src):
+                assumed that the attributes are Image/Frame level attrs.
+            median_point (float): value between [0, 1] to specify what the
+                target count per attribute value will be.
+                0.5 - will result in the true median
+                0   - the minimum value
+                It is recommended to set this somewhere between [0, 0.5]. The
+                smaller this value is, the closer all values can be balanced,
+                at the risk that if some values have particularly low number of
+                samples, they dataset will be excessively trimmed.
+            negative_power (float): value between [1, ~LARGE~] that weights the
+                negative values (where the count of a value is less that the
+                target) when computing the score for a set of indices to remove.
+                1 - will weight them the same as positive values
+                2 - will square the values
+                ...
+                Check Balancer._solution_score for more details.
+        '''
+        self.attr_name = attribute_name
+        self.object_label = object_label
+        self.median_point = median_point
+        self.negative_power = negative_power
 
     def transform(self, src):
-        # @TODO implement Balancing!!
+        '''
+        Modify the BuilderDataset records by removing records until the
+        target attribute is ~roughly~ balanced for each value.
+
+        Args:
+            src (BuilderDataset): the dataset builder
+        '''
+
+        # STEP 1: Get attribute value(s) for every record
+        helper_list = self._to_helper_list(src.records)
+
+        # STEP 2: determine target number to remove of each attribute value
+        target_remove = self._get_target_remove(helper_list)
+
+        # STEP 3: find the records to keep
+        keep_idxs = self._get_keep_idxs(target_remove, helper_list)
+
+        # STEP 4: modify the list of records
         old_records = src.records
         src.clear()
-        pass
+        for ki in keep_idxs:
+            src.add(old_records[helper_list[ki][0]])
 
+    def _to_helper_list(self, records):
+        '''
+
+        records: list of BuilderDataRecord's
+        return: a list of tuples with two entries:
+
+                [
+                    (record_id, list_of_values),
+                    (record_id, list_of_values),
+                    (record_id, list_of_values),
+                    ...
+                ]
+
+            record_id: integer ID of the corresponding old_record
+            list_of_values: list of attribute values for the attribute to be
+                            balanced, one per unique object, if using objects.
+                            For example: ['red', 'red', 'green'] would imply
+                            three objects with the 'color' attribute in this
+                            record.
+        '''
+        helper_list = []
+
+        if not len(records):
```

Comment:
this is a nit point for me. I think that the way I wrote it helps emphasize what my intention was. If this is a strict ETA rule, then I'll change it, but otherwise I think that adding `len(...)` makes it clear that I am not check for `None` or `False` or some other truthy term.

Reply:
Whether it's a strict ETA rule is up to all of us!  I think it is mentioned in the style guide somewhere, but I don't think this way is so bad.  What I would object to more is something like `if len(records) == 0:`.

Reply:
Agreed, to me, `len(records)` without `== 0` is both concise and clear. So I'm going to resolve this.